### ISSUE TITLE
incompatible List api

### ISSUE THREAD

**Issue by avdosev** [dart-lang/sdk#48829]
The list has a remove method with the return value of a flag indicating whether the item has been removed from the list

But removeWhere and retainWhere return void

It would be nice if these methods return the same as remove

**Comment by lrhn** [comment]
Nice idea, but too breaking to change the `List` API at this time, at least for something this small.
If we ever do a breaking change to `List` (definitely won't happen unless we get an API versioning feature first), then it's something we can consider.

Absent that, it's not going to happen. (Sadly, it's not a bad idea otherwise.)

**Comment by unterumarmung** [comment]
I was just passing by and saw this issue.
@lrhn, could you please explain why changing the return type from void to non-void is a breaking change? It is really not obvious.

**Comment by julemand101** [comment]
> I was just passing by and saw this issue. @lrhn, could you please explain why changing the return type from void to non-void is a breaking change? It is really not obvious.

You can create your own list implementation that implements the `List` interface (Dart already does this for several types of lists). If you change the return type of some of the interface methods, all implementations will break since the need to make the change.

**Comment by lrhn** [comment]
Yep, that's the reason. It could invalidate other classes implementing `List` and still returning `void`.

Almost all changes to a Dart class API is potentially breaking.

If you change a return type to be more specific, like from `num` to `int`, any class implementing that interface now becomes invalid. Same if adding an optional parameter, or changing a required parameter to optional, or changing a parameter type to be more general. Or adding an instance member.
If you change a return type to be more general, like from `int` to `num`, any user of that member may now be invalid (they get a `num` where an `int` is needed). Same if removing a parameter, or changing an optional parameter to required, or changing a parameter type to be more specific. Or removing an instance member.

Adding a static member to a class is one of the last non-breaking API changes available. Even extension members can conflict with other extension members.
